// @verdaccio/search-indexer@8.0.2/build/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let debug = require("debug");
debug = __toESM(debug);
let fuse_js = require("fuse.js");
fuse_js = __toESM(fuse_js);
//#region src/indexer.ts
var debug$1 = (0, debug.default)("verdaccio:search:indexer");
var FUSE_OPTIONS = {
	keys: [
		{
			name: "name",
			weight: 2
		},
		{
			name: "keywords",
			weight: 1.2
		},
		{
			name: "description",
			weight: 1
		},
		{
			name: "author",
			weight: .5
		}
	],
	includeScore: true,
	threshold: .4,
	ignoreLocation: true
};
var SearchMemoryIndexer = class {
	fuse;
	storage;
	logger;
	configureStorage(storage) {
		this.storage = storage;
	}
	async query(term) {
		if (!this.fuse) return { hits: [] };
		debug$1("searching %s at indexer", term);
		return { hits: this.fuse.search(term).map((r) => ({
			...r.item,
			score: r.score
		})) };
	}
	prepareKeywords(keywords) {
		if (typeof keywords === "undefined") return "";
		else if (typeof keywords === "string") return keywords;
		return keywords.join(",");
	}
	async add(pkg) {
		if (!this.fuse) return;
		const name = pkg.name;
		debug$1("adding item %s to the indexer", name);
		const item = {
			id: name,
			name,
			description: pkg.description ?? "",
			version: pkg.version,
			keywords: this.prepareKeywords(pkg.keywords),
			author: pkg._npmUser ? pkg._npmUser.name : ""
		};
		this.fuse.remove((doc) => doc.id === name);
		this.fuse.add(item);
	}
	async remove(name) {
		if (!this.fuse) return;
		debug$1("removing item %s from the indexer", name);
		this.fuse.remove((doc) => doc.id === name);
	}
	async reindex() {
		debug$1("reindexing search indexer");
		this.storage?.getLocalDatabase(async (error, packages) => {
			if (error) throw error;
			let i = packages.length;
			if (i === 0) debug$1("no packages to index");
			while (i--) {
				const pkg = packages[i];
				debug$1("indexing package %s", pkg?.name);
				try {
					await this.add(pkg);
				} catch (err) {
					this.logger?.error({ err: err.message }, "error @{err} indexing package");
				}
			}
			debug$1("reindexed search indexer");
		});
	}
	async init(logger) {
		this.logger = logger;
		this.fuse = new fuse_js.default([], FUSE_OPTIONS);
		this.reindex();
	}
};
var indexer_default = new SearchMemoryIndexer();
//#endregion
exports.SearchMemoryIndexer = indexer_default;
